### PR TITLE
Cache DataFile Constructor method

### PR DESCRIPTION
**Pull request by LuciferYang**
## What changes were proposed in this pull request?
As #554 said, add a constructor method cache to avoid reflect every times 

## How was this patch tested?
DataFileSuite

### CODE REVIEW

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  override def afterEach(): Unit = {}
+
+  test("apply and cache") {
+    val data = (0 to 10).map(i => (i, (i + 'a').toChar.toString))
+    val schema = new StructType()
+    val config = new Configuration()
+
+
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala]: Nit: 1 empty line is enough

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.Utils
+
+class DataFileSuite extends QueryTest with SharedOapContext {
+
+  override def beforeEach(): Unit = {
+    val path = Utils.createTempDir().getAbsolutePath
+  }
+
+  override def afterEach(): Unit = {}
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala]: why not call `super.afterEach()`?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala (hunk 1 above)
Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala]: ok, I'll remove it.

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala (hunk 2 above)
Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala]: DataFile use createDataFileHandle method to init val meta and there is no method to force release meta resource from cache, if call `super.afterEach()` will throw a exception from DebugFileSystem `There are 1 possibly leaked file streams.`, any suggestion ? @adrian-wang

**Comment by LuciferYang** [comment]
@adrian-wang use guava cache because of it can init cache item from a cacheloader and it threadsafe. Use a simple map if we add a new DataFile implementation like OrcDataFile, we need add new code to manual put new constuctor to map, if we accept this cost I think we should use `case className => new XXXDataFile(xxx)` to instead of reflection api.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFile.scala [hunk 3]
```diff
@@ -50,10 +53,16 @@ private[oap] class OapIterator[T](inner: Iterator[T]) extends Iterator[T] with A
 }
 
 private[oap] object DataFile {
+
+  private val cache = CacheBuilder.newBuilder().build(new CacheLoader[String, Constructor[_]] {
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFile.scala]: @chenghao-intel, I'll fix it~  declare cache to LoadingCache[String, Constructor[_]]

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFile.scala [hunk 4]
```diff
@@ -65,6 +74,8 @@ private[oap] object DataFile {
           s" (String, StructType) for class $dataFileClassName")
     }
   }
+
+  private[oap] def cacheSize: Long = cache.size()
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFile.scala]: @chenghao-intel This method only use by DataFileSuite to assertion cacheSize max value is 2, DataFile. cache is private, don't expose this we cannot but reflection DataFile. cache value :(, Any suggestion?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    assert(DataFile.cacheSize == 2)
+
+    intercept[OapException] {
+      DataFile("nofile", schema, "NotExistClass", config)
+      assert(DataFile.cacheSize == 2)
+    }
+  }
+
+  private def listDirectory(path: File): util.List[String] = {
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala]: Maybe we can use SpecificParquetRecordReaderBase.listDirectory api directly ~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFile.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFile.scala]: @chenghao-intel Now I rename it to cachedConstructorCount , or we shlould get this value use reflection api in test case and remove this method ?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala (hunk 2 above)
Comment (adrian-wang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala]: I think we'd better add a comment here explaining that.

Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala]: ok~ I'll add a comment  today

Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFileSuite.scala]: Thx~ @adrian-wang  Already add comment to explain why override aftereach method in DataFileSuite

**Comment by LuciferYang** [comment]
@adrian-wang fix confilcts ~

**Comment by adrian-wang** [comment]
merging this, thanks @LuciferYang

**Comment by LuciferYang** [comment]
Thx~ @adrian-wang

**Pull request merged**